kernel: userspace: dynamic thread stack support

ceolin opened the pull request:
Support for dynamic thread stack on userspace.

Commit:
kernel: userspace: Add k_object_alloc_size

Add a new API to dynamically allocate kernel objects that allow
passing an arbitrary size. This new API allows to allocate dynamic
thread stack.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
kernel: dynamic: Support usermode thread stack

Allocate kernel object for userspace thread stack.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
kernel: userspace: Dynamic thread stack object

Add support for dynamic thread stack objects. A new container
for this kernel object was added to avoid its alignment constraint
to all dynamic objects.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

cfriedt commented:
@ceolin - I've applied your fixups and pushed to `issue/26999/k-thread-create-dynamic-stack`

Commit:
kernel: dynamic: Fix stack allocation logic

Fix the preference allocation logic. If pool is preferred but POOL_SIZE
is 0 or pool allocation fails, it fallbacks to heap allocation if it
is enabled.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
tests: kernel: threads: stack: Enable USERSPACE

Test that automatic thread stack allocation works for
both user and kernel threads.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
kernel: userspace: Use only list for dynamic objs

Since the rbtree is using as list because we no longer
can assume that the object pointer is the address of the
data field in the dynamic object struct, lets just use
the already existent dlist for tracking dynamic kernel
objects.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

cfriedt commented on kernel/userspace.c:
```diff
@@ -192,25 +191,14 @@ extern struct z_object *z_object_gperf_find(const void *obj);
 /*
- * TODO: Write some hash table code that will replace both obj_rb_tree
- * and obj_list.
+ * TODO: Write some hash table code that will replace obj_list.
```
There is a hash table implementation if you are looking for one [hash_map_api.h](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/sys/hash_map_api.h)

andyross reviewed (commented):
I don't see anything wrong, but that wasted page for every stack creeps me out a bit.  Seems like one extra layer of indirection is needed.

ceolin commented on kernel/userspace.c:
```diff
@@ -130,41 +149,56 @@ uint8_t *z_priv_stack_find(k_thread_stack_t *stack)
+	struct dyn_obj_base base;
+
+	/* The object itself */
+	uint8_t data[] __aligned(DYN_OBJ_DATA_ALIGN_K_THREAD_STACK);
```
yep ... there is this waste of memory indeed, that is probably no longer needed now that I am searching the object linearly. I can now do two allocations and avoid it. Let met try it out.

ceolin replied:
I removed this alignment requirement for the object and make a new allocation for the stack data itself (with the proper alignment), this can potentially avoid some memory waste.

ceolin commented on kernel/userspace.c:
```diff
@@ -193,6 +227,9 @@ static size_t obj_align_get(enum k_objects otype)
 #endif
 		break;
+	case K_OBJ_THREAD_STACK_ELEMENT:
+		ret = __alignof(struct dyn_obj_stack);
```
that function is asking for the object alignment and not the data itself.

andyross commented on kernel/userspace.c:
```diff
@@ -130,41 +149,56 @@ uint8_t *z_priv_stack_find(k_thread_stack_t *stack)
+	struct dyn_obj_base base;
 
 	/* The object itself */
 	uint8_t data[] __aligned(DYN_OBJ_DATA_ALIGN_K_THREAD);
```
Shouldn't this item be removed now?  The address where it would be is now (in a dyn_obj_stack) the "data" pointer (identically named, but a pointer and not an array!).  Trying to use this now would be an aliasing bug.

ceolin replied:
I think I can make everything work with only one object type now. This was needed to avoid imposing the stack alignment requirement for all objects. I can try it out later.

andyross reviewed (approved):
One more nitpick about the confusingly aliased "data" fields.  FWIW I'm seeing now where that alignment padding in the struct came from, it wasn't originally for page padding, it was to get the struct k_thread aligned correctly on x86, which (IIRC) needs that for the x87/SSE context switch field.

No reason not to approve, but if the two uses of data became a proper union with an assert or something to track misuse, I wouldn't cry.

edersondisouza commented on kernel/userspace.c:
```diff
@@ -348,7 +348,8 @@ void *z_impl_k_object_alloc(enum k_objects otype)
 	/* The following are currently not allowed at all */
```
This comment ceases to make sense, right?
